src/ripple/protocol/impl/STAmount.cpp

```diff
@@ -980,6 +980,58 @@ operator- (STAmount const& value)
 // Arithmetic
 //
 //------------------------------------------------------------------------------
+
+// Calculate (a * b) / c when all three values are 64-bit
+// without loss of precision:
+static
+std::uint64_t
+fast_muldiv(
+    std::uint64_t a,
+    std::uint64_t b,
+    std::uint64_t c)
+{
+    boost::multiprecision::uint128_t ret;
+
+    boost::multiprecision::multiply(ret, a, b);
+    ret /= c;
+
+    if (ret > std::numeric_limits<std::uint64_t>::max())
+    {
+        throw std::overflow_error ("overflow: " +
+            std::to_string (a) + " * " +
+            std::to_string (b) + " / " +
+            std::to_string (c));
+    }
+
+    return static_cast<uint64_t>(ret);
+}
+
+static
+std::uint64_t
+fast_muldiv(
```

Comment:
Could use a comment like the one line 984.  I'm also not sold on the function name, since `d` is added.  Perhaps `mulAddDiv()` in which case `c` should be added and `d` divided.  Finally, I'm guessing that the primary point of this function and the preceding is to be lossless more than fast.  So perhaps `lossless_mulDiv()` and `lossless_mulAddDiv()`?  Those names are probably too long.  I'm guessing you can come up with something better, but you get the idea.

I don't see any functional problems here, just thinking about naming.  Your call on which of these ideas you want to use, if any.